xls/modules/zstd/buffer.x

```diff
@@ -0,0 +1,451 @@
+        (BufferStatus::NO_ENOUGH_DATA, bits[BSIZE]:0)
+    } else {
+        let mask = (bits[BSIZE]:1 << length) - bits[BSIZE]:1;
+        (BufferStatus::OK, buffer.contents & mask)
+    }
+}
+
+#[test]
+fn test_buffer_peek_checked() {
+    let buffer = Buffer { contents: u32:0xDEADBEEF, length: u32:32 };
+
+    let (status1, data1) = buffer_peek_checked(buffer, u32:0);
+    assert_eq(status1, BufferStatus::OK);
+    assert_eq(data1, u32:0);
+
+    let (status2, data2) = buffer_peek_checked(buffer, u32:16);
+    assert_eq(status2, BufferStatus::OK);
+    assert_eq(data2, u32:0xBEEF);
+
+    let (status3, data3) = buffer_peek_checked(buffer, u32:32);
+    assert_eq(status3, BufferStatus::OK);
+    assert_eq(data3, u32:0xDEADBEEF);
+
+    let (status4, data4) = buffer_peek_checked(buffer, u32:64);
+    assert_eq(status4, BufferStatus::NO_ENOUGH_DATA);
+    assert_eq(data4, u32:0);
+}
+
+// Creates a new buffer
+pub fn buffer_new<BSIZE: u32>() -> Buffer<BSIZE> {
+    Buffer { contents: bits[BSIZE]:0, length: u32:0 }
+}
+
+#[test]
+fn test_buffer_new() {
+    assert_eq(buffer_new<u32:32>(), Buffer { contents: u32:0, length: u32:0 });
+}
+
+// Example use case
+//
+// The Buffer structure can be used to collect enough data that is required
+// for the next steps to proceed. The next steps may consume variable or
+// constant amount of data. In the simplest use-case scenario, one can use
+// it to combine smaller transactions into bigger ones.
+
+type Buffer64 = Buffer<64>;
+
+proc BufferExampleUsage {
+    input_r: chan<u32> in;
+    output_s: chan<u48> out;
+
+    config(
+        input_r: chan<u32> in,
+        output_s: chan<u48> out
+    ) { (input_r, output_s) }
+
+    init { Buffer64 { contents: u64:0, length: u32:0 } }
+
+    next(tok: token, buffer: Buffer64) {
+        let (tok, recv_data) = recv(tok, input_r);
+        let buffer = buffer_append<u32:64>(buffer, recv_data);
+
+        if buffer.length >= u32:48 {
+            let (buffer, data_to_send) = buffer_sized_pop<u32:64, u32:48>(buffer);
+            let tok = send(tok, output_s, data_to_send);
+            buffer
+        } else {
+            buffer
+        }
+    }
+}
+
+const NUMBER_OF_TESTED_TRANSACTIONS = u32:30;
+
+struct BufferExampleUsageTestState {
+    recv_count: u32,
+    send_count: u32
+}
+
+#[test_proc]
+proc BufferExampleUsageTest {
+    terminator: chan<bool> out;
+    data32_s: chan<u32> out;
+    data48_r: chan<u48> in;
+
+    config(terminator: chan<bool> out) {
+        let (data32_s, data32_r) = chan<u32>;
+        let (data48_s, data48_r) = chan<u48>;
+        spawn BufferExampleUsage(data32_r, data48_s);
+        (terminator, data32_s, data48_r)
+    }
+
+    init {
+        BufferExampleUsageTestState {
+            recv_count: u32:0,
+            send_count: u32:0
+        }
+    }
+
+    next(tok: token, state: BufferExampleUsageTestState) {
+        let data_to_send = match state.send_count % u32:3 {
```

Comment:
curious if you've tried sequencial send from a single next tick of the `test_proc` instead? that should be more readable.

Reply:
Indeed! It is much simpler with this approach. Thanks!